docs/adr-custom-exceptions.md

```diff
@@ -0,0 +1,23 @@
+# InstructLab Custom Exceptions Pattern
+
+## Context
+
+The historically pervasive pattern for exception handling during application runtime is to catch internally raised exceptions in the CLI layer and use `click.secho` followed by `click.exceptions.Exit` to display a useful error message to the user before exiting the application. This leaves a risk of intermediate calls between the site of the exception and the user-facing layer changing, leading to missed new exceptions and outdated caught exceptions. A second issue is that of discoverability and verification: given a `click.exceptions.Exit` exception handling, it is not clear from the code where the caught exception originates from in the call stack, and, similarly, given a piece of code that can raise an exception, it is not clear from the local code whether that exception is properly handled in the CLI layer without investigation.
+
+These issues will compound whenever REST APIs begin development.
```

Comment:
and/or whenever the [SDK APIs](https://github.com/instructlab/dev-docs/pull/184) begin development.